server/src/main/java/org/elasticsearch/plugins/UberModuleClassLoader.java

```diff
@@ -0,0 +1,281 @@
+    static UberModuleClassLoader getInstance(ClassLoader parent, String moduleName, List<Path> jarPaths, Set<String> moduleDenyList) {
+        ModuleFinder finder = ModuleSupport.ofSyntheticPluginModule(moduleName, jarPaths.toArray(new Path[0]), Set.of());
+        List<URL> jarURLs = new ArrayList<>();
+        try {
+            for (Path jarPath : jarPaths) {
+                URI toUri = jarPath.toUri();
+                URL toURL = toUri.toURL();
+                jarURLs.add(toURL);
+            }
+        } catch (IOException e) {
+            throw new IllegalArgumentException(e);
+        }
+        ModuleLayer mparent = ModuleLayer.boot();
+        Configuration cf = mparent.configuration().resolve(finder, ModuleFinder.of(), Set.of(moduleName));
+
+        Set<String> packageNames = new HashSet<>();
+        for (URL url : jarURLs) {
+            try (JarFile jarFile = new JarFile(new File(url.toURI()))) {
+                Set<String> jarPackages = ModuleSupport.scan(jarFile)
+                    .classFiles()
+                    .stream()
+                    .map(e -> ModuleSupport.toPackageName(e, "/"))
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(Collectors.toSet());
+
+                packageNames.addAll(jarPackages);
+            } catch (IOException | URISyntaxException e) {
+                throw new IllegalArgumentException(e);
+            }
+        }
+
+        PrivilegedAction<UberModuleClassLoader> pa = () -> new UberModuleClassLoader(
+            parent,
+            moduleName,
+            jarURLs.toArray(new URL[0]),
+            cf,
+            mparent,
+            moduleDenyList,
+            packageNames
+        );
+        return AccessController.doPrivileged(pa);
+    }
+
+    /**
+     * Constructor
+     */
+    private UberModuleClassLoader(
+        ClassLoader parent,
+        String moduleName,
+        URL[] jarURLs,
+        Configuration cf,
+        ModuleLayer mparent,
+        Set<String> moduleDenyList,
+        Set<String> packageNames
+    ) {
+        super(parent);
+
+        this.internalLoader = new URLClassLoader(jarURLs);
+        // code source is always the first jar on the list
+        this.codeSource = new CodeSource(jarURLs[0], (CodeSigner[]) null);
+        // we need a module layer to bind our module to this classloader
+        this.moduleController = ModuleLayer.defineModules(cf, List.of(mparent), s -> this);
+        this.module = this.moduleController.layer().findModule(moduleName).orElseThrow();
+
+        // Every module reads java.base by default, but we can add all other modules
+        // that are not in the deny list so that plugins can use, for example, java.management
+        mparent.modules()
+            .stream()
+            .filter(Module::isNamed)
+            .filter(m -> "java.base".equals(m.getName()) == false)
+            .filter(m -> moduleDenyList.contains(m.getName()) == false)
```

Comment:
deny-list is great, but at some point it might be more straightforward to have an allow-list, say to allow readability to all the JDK modules and a subset of the ES modules.